kernel/src/utilities/streaming_process_slice.rs

```diff
@@ -0,0 +1,269 @@
+///    space left, the kernel will set the `exceeded` flag bit (index 0).
+///
+///    The `halt` flag bit as set by the process governs the kernel's behavior
+///    once the `exceeded` flag is set: if `halt` is cleared, the kernel will
+///    attempt to write future, smaller chunks to be buffer (and thus implicitly
+///    discarding some packets). If `halt` and `exceeded` are both set, the
+///    kernel will stop writing any data into the buffer.
+///
+/// 5. The kernel will schedule an upcall to the process, indicating that a
+///    write to the buffer (or setting the `exceeded`) flag occurred. The kernel
+///    may schedule only one upcall for the first chunk written to the buffer,
+///    or multiple upcalls (e.g., one upcall per chunk written). A process must
+///    not rely on the number of upcalls received and instead rely on the buffer
+///    metadata (`offset` and the `flags` bits) to determine the amount of data
+///    written to the buffer.
+///
+/// 6. The process prepares its second buffer, following step 2. The process
+///    then issues an `allow` operation that atomically swaps the current
+///    allowed buffer by its second buffer.
+///
+/// 7. The process can now process the received chunks contained in the initial
+///    buffer, while the kernel receives new chunks in the other, newly allowed
+///    buffer.
+///
+///
+/// The version 0 buffer format is specified as follows:
+/// ```text,ignore
+/// 0          1          2          3          4          5
+/// +----------+----------+----------+----------+----------+-------------------...
+/// | flags    | buffer offset                             | data
+/// +----------+----------+----------+----------+----------+-------------------...
+/// | VVVVxxHE | 32 bits (native endian)                   |
+/// ```
+///
+/// The flags field is structured as follows:
+/// - `V`: version bits. This kernel only supports version `0`.
+/// - `H`: `halt` flag. If this flag is set alongside the `exceeded` flag, the
+///   kernel will not write any further data to this buffer.
+/// - `E`: `exceeded` flag. The kernel sets this flag when the remaining buffer
+///   capacity is insufficient to append the current chunk.
+/// - `x`: reserved flags. Unless specified otherwise, processes must clear
+///   these flags prior to `allow`ing a buffer to the kernel. A kernel that does
+///   not know of a reserved flag must refuse to operate on a buffer that has
+///   such a flag set.
+#[repr(transparent)]
+pub struct StreamingProcessSlice<'a> {
+    slice: &'a WriteableProcessSlice,
+}
+
+register_bitfields![
+    u8,
+    pub StreamingProcessSliceFlags [
+        VERSION OFFSET(4) NUMBITS(4) [
+            V0 = 0x00,
+        ],
+        RESERVED OFFSET(2) NUMBITS(2) [
+            RESERVED0 = 0x00,
+        ],
+        HALT OFFSET(1) NUMBITS(1) [],
+        EXCEEDED OFFSET(1) NUMBITS(1) [],
+    ]
+];
+
+#[derive(Debug, Clone, Copy)]
+pub struct StreamingProcessSliceMeta {
```

Comment:
```suggestion
pub struct StreamingProcessSliceHeader {
```

Reply:
I think that's a good change, I'll make sure to change it everywhere consistently.